File: app/src/types/Item.ts
import type { Shape } from '@miwos/shape'
import type { Point, Signal } from '.'

export type ItemCategory = 'modules' | 'modulators'

export interface Item {
  id: number
  category: ItemCategory
  type: string
  label: string
  props: Record<string, unknown>
  modulatedProps: Record<string, unknown>
  position: Point
  inputs?: { signal: Signal; position?: Point; angle?: number }[]
  outputs?: { signal: Signal; position?: Point; angle?: number }[]
}

export interface ItemSerialized {
  id: number
  category: ItemCategory
  type: string
  label?: string
  props?: Record<string, unknown>
  position?: [x: number, y: number]
}

export interface ItemDefinition {
  id: string
  category: ItemCategory
  label?: string
  props: Record<string, { type: string; options: Record<string, any> }>
  inputs: { signal: Signal; position?: Point; angle?: number }[]
  outputs: { signal: Signal; position?: Point; angle?: number }[]
  shape?: Shape['id']
}

export interface ItemDefinitionSerialized {
  id: string
  category: ItemCategory
  label?: string
  props?: Record<string, [type: string, options: Record<string, any>]>
  inputs?: Signal[]
  outputs?: Signal[]
  shape?: Shape['id']
}

File: engine/src/types/Item.lua
---@meta

---@alias ItemCategory 'modules' | 'modulators'

---@class ItemSerialzied
---@field id number
---@field type string
---@field props table<string, any>

---@class ItemDefinition
---@field category ItemCategory
---@field props table<string, Prop>
---@field inputs Signal[]?
---@field outputs Signal[]?

---@class ItemDefinitionSerialized
---@field category ItemCategory
---@field props table<string, { [1]: string, [2]: table<string, any> }>
---@field inputs Signal[]?
---@field outputs Signal[]?
